File: frontend/src/api/services/message.service.ts
import api from '../axios'
import type { Message, CreateMessageDTO } from '../types'

class MessageService {
    /**
     * Get all messages
     */
    async getAllMessages(): Promise<Message[]> {
        const response = await api.get<Message[]>('/messages')
        return response.data
    }

    /**
     * Get message by ID
     */
    async getMessageById(id: number): Promise<Message> {
        const response = await api.get<Message>(`/messages/${id}`)
        return response.data
    }

    /**
     * Get messages by conversation
     */
    async getMessagesByConversation(conversationId: number): Promise<Message[]> {
        const response = await api.get<Message[]>(`/messages/conversation/${conversationId}`)
        return response.data
    }

    /**
     * Get messages by user
     */
    async getMessagesByUser(userId: number): Promise<Message[]> {
        const response = await api.get<Message[]>(`/messages/user/${userId}`)
        return response.data
    }

    /**
     * Create new message
     */
    async createMessage(message: CreateMessageDTO): Promise<Message> {
        const response = await api.post<Message>('/messages', message)
        return response.data
    }

    /**
     * Update message
     */
    async updateMessage(id: number, content: string): Promise<Message> {
        const response = await api.put<Message>(`/messages/${id}`, { content })
        return response.data
    }

    /**
     * Delete message
     */
    async deleteMessage(id: number): Promise<void> {
        await api.delete(`/messages/${id}`)
    }
}

export const messageService = new MessageService()

File: backend/src/main/java/com/tictac/tictac/service/MessageService.java
package com.tictac.tictac.service;

import java.util.List;
import java.util.Optional;

import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.tictac.tictac.entity.Conversation;
import com.tictac.tictac.entity.Message;
import com.tictac.tictac.event.MessagePostedEvent;
import com.tictac.tictac.repository.ConversationRepository;
import com.tictac.tictac.repository.MessageRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
@Transactional
public class MessageService {

    private final MessageRepository messageRepository;
    private final ConversationRepository conversationRepository;
    private final ApplicationEventPublisher events;

    public List<Message> getAllMessages() {
        return messageRepository.findAll();
    }

    public Optional<Message> getMessageById(Long id) {
        return messageRepository.findById(id);
    }

    public List<Message> getMessagesByConversation(Long conversationId) {
        Conversation conversation = conversationRepository.findById(conversationId)
                .orElseThrow(() -> new RuntimeException("Conversation not found with id: " + conversationId));
        return messageRepository.findByConversation(conversation);
    }

    public List<Message> getMessagesByUser(Long userId) {
        return messageRepository.findByAuthorIdUser(userId);
    }

    public Message createMessage(Message message) {
        Message saved = messageRepository.save(message);
        events.publishEvent(new MessagePostedEvent(saved));
        return saved;
    }

    public Message updateMessage(Long id, Message messageDetails) {
        return messageRepository.findById(id)
                .map(message -> {
                    message.setContent(messageDetails.getContent());
                    return messageRepository.save(message);
                })
                .orElseThrow(() -> new RuntimeException("Message not found with id: " + id));
    }

    public void deleteMessage(Long id) {
        messageRepository.deleteById(id);
    }
}
